### ISSUE TITLE
Testing DSpot in Atos Supersede IF Project: All tests did not pass without mutation when calculating line coverage. Mutation testing requires a green suite.

### ISSUE THREAD

**Issue by jesus-gorronogoitia** [STAMP-project/dspot#240]
**Issue Type**: possible bug/misuse
**Description**: I've tested DSpot on [Atos Supersede IF project](https://github.com/supersede-project/integration/tree/descartes/IF/API/eu.supersede.if.api). Maven passes the test suite declared to be included, without errors. However, when tested in DSpot, it complains the test suite is not green. Even executed with --verbose, I get not information about what concrete test case (in the suite) failed. See Steps to reproduce for more details. 
**Reproducibility**: always
**Severity**: major
**Tool**: Descartes PITest, version 25/10/17
**Execution Environment**: Linux OpenSuse Tumbleweed
**Steps to reproduce**:
Clone Atos Supersede IF project from GitHub (descartes branch):
`git clone https://github.com/supersede-project/integration.git -b descartes`
Go to IF folder:
`cd integration/IF/API/eu.supersede.if.api`
You can run one selected TestSuite, declared in pom_for_dspot.xml, namely ReplanControllerProxyTest.java, by using Maven:
mvn -f pom_for_dspot.xml test
so only this test is executed by Maven. It passes correctly.
This is the pom_dspot.xml configuration for filtering the test suite to be executed:
`<plugin>
		        <groupId>org.apache.maven.plugins</groupId>
		        <artifactId>maven-surefire-plugin</artifactId>
		        <version>2.20</version>
		        <configuration>
		          <includes>
		            <include>**/ReplanControllerProxyTest.java</include>
		          </includes>
		        </configuration>
		    </plugin>`
Unlike issue 28 for Descartes, no system parameters are required for this test suite.
However, if I launch DSpot, using: 
`java -jar target/dspot-1.0.0-jar-with-dependencies.jar  -p ../../../Supersede/Git/integration/IF/API/eu.supersede.if.api/dspot.properties -i 1 -t eu.supersede.integration.api.replan.controller.proxies.test.ReplanControllerProxyTest -a MethodAdd`
DSpot complains that

> 10:30:32 AM PIT >> SEVERE : Description [testClass=eu.supersede.integration.api.replan.controller.proxies.test.ReplanControllerProxyTest, name=eu.supersede.integration.api.replan.controller.proxies.test.ReplanControllerProxyTest] did not pass without mutation.

> [ERROR] Failed to execute goal org.pitest:pitest-maven:1.2.2:mutationCoverage (default-cli) on project eu.supersede.if.api: Execution default-cli of goal org.pitest:pitest-maven:1.2.2:mutationCoverage failed: All tests did not pass without mutation when calculating line coverage. Mutation testing requires a green suite.

I don't know which concrete test within ReplanControllerProxyTest fail, so I can investigate further.

Next is the dspot.properties used for this experimentation:

```
#relative path to the project root from dspot project
project=../../../Supersede/Git/integration/IF/API/eu.supersede.if.api
#relative path to the source project from the project properties
src=src/main/java/
#relative path to the test source project from the project properties
testSrc=src/test/java
#java version used
javaVersion=8
# (optional) path to the output folder, default to "output_diversify"
outputDirectory=dspot-out/
# (optional) filter on the package name containing tests to be amplified ("example" => "example.*"
filter=eu.supersede.integration.api.replan.controller.proxies.test*

```
**Files**:

**Relationships**: Same as issue number 28 for Descartes
**Reporter**: Yosu Gorroñogoitia jesus.gorronogoitia@atos.net

**Comment by danglotb** [comment]
Hello @jesus-gorronogoitia 

Thank you for taking the time to use **DSpot**.

After investigation, I figured out that you are using some properties file in `src/main/resources/`.

In fact, for **DSpot**, you have to explicitly specify the path of resources.

I suggest you to add the following line to your property file:
```java
srcResources=src/main/resources/
 ```

I am sorry that the documentation is not enough clear, and most important,  out of date.

I think I am gonna make some default location inside **DSpot** (such as `src/main/resource/` and `src/test/resources/`), what do you think?

Thank again.

**Comment by danglotb** [comment]
I just see that there is still an error. I'am investigating.

**Comment by jesus-gorronogoitia** [comment]
Merci Benjamin,
I will follow your advice asap. I am fine with your proposal to support default location for Maven standard project locations such as those you mention.

**Comment by danglotb** [comment]
Okay, It is because **DSpot** is trying to run **pit** using the whole test suite.

By default **DSpot** uses the *PitMutantScoreSelector* which keeps amplified test that increase the mutation score. To do this, it must first compute the mutation score of the original test suite.

So, **DSpot** compute the mutation score using the whole test suite by running the following goal maven:
`org.pitest:pitest-maven:mutationCoverage -DwithHistory -DtargetClasses="eu.supersede.integration.api.*" -DreportsDirectory=target/pit-reports -DoutputFormats=CSV,HTML -DtimeoutConst=10000 -DjvmArgs=16G -DmutationEngines=gregor -Dmutators=ALL -DadditionalClasspathElements=target/dspot/dependencies/ `

without no values for `-DtargetClasses`, which it will use all available test.

That results by running none passing tests, which make **DSpot** fail.

There is two solutions to this:

1. Pre-compute the mutation score with for instance (from the root of `eu.supersede.if.api`):
`sh
mvn clean test -DskipTests org.pitest:pitest-maven:mutationCoverage -DwithHistory -DtargetClasses="eu.supersede.integration.api.*" -DreportsDirectory=target/pit-reports -DoutputFormats=CSV -DtimeoutConst=10000 -DjvmArgs=16G -DmutationEngines=gregor -Dmutaorrs=ALL -DadditionalClasspathElements=target/dspot/dependencies/ -DtargeTests=eu.supersede.integration.api.replan.controller.proxies.test.ReplanControllerProxyTest
`
 to obtain a `mutations.csv` and feed **DSpot** with it using the `-m` command line (`-m <path-to-mutation-reports.csv>`.

2. Add to your property file, the exhaustive list of failing tests using the property
`excludedClasses`. Example: 

```java
excludedClasses=my.class.TestClass,my.package.*
```

**BUT** It seems that `ReplanControllerProxyTest` takes a lot of time to be executed. Such test will make the mutation analysis, and **DSpot**'s execution very long.

In fact, the mutation analysis required to run the test a lot of time against different variants/mutants of the program.

I have only one suggestion to "overcome" this issue: do not use *PitMutantScoreSelector*.

That is to say, I think it will be better, for sake of time to use another selector, such as the *JacocoCoverageSelector*, which keeps only amplified test that increase the instruction coverage. I think this is weaker that the mutation score, but much (much ++) more faster.

To do this, you can use the command line option: `-s JacocoCoverageSelector` or `--test-criterion JacocoCoverageSelector`

Another problem that I found, is that you use System Properties within the source (as you showed in STAMP-project/pitest-descartes#28 )

To solve this, you can pass those System Properties throught the command line: `-Dis.admin.user=admin -Dis.admin.passwd=$2pRSid#` 
Full command-line:

`java -jar -Dis.admin.user=admin -Dis.admin.passwd=$2pRSid# target/dspot-1.0.0-jar-with-dependencies.jar -p ../integration/IF/API/eu.supersede.if.api/dspot.properties -i 1 -t eu.supersede.integration.api.replan.controller.proxies.test.ReplanControllerProxyTest -a MethodAdd --test-criterion JacocoCoverageSelector`


Last thing, since your test is taking a while to be executed, you will have a `java.util.concurrent.TimeoutException` and **DSpot** will stops brutally. 

I will open a pull request, with the possibility to tune the time to wait until  `java.util.concurrent.TimeoutException` is thrown.

I will tell you when the changes are ready.

Thank you very much! :)

**Comment by danglotb** [comment]
Hello, I just merged #244

I ran **DSpot** on [Atos Supersede IF project](https://github.com/supersede-project/integration/tree/descartes/IF/API/eu.supersede.if.api) with:
`-p ../integration/IF/API/eu.supersede.if.api/dspot.properties -i 1 -t eu.supersede.integration.api.replan.controller.proxies.test.ReplanControllerProxyTest -a MethodAdd --test-criterion JacocoCoverageSelector` options.

My results are:

>======= REPORT =======
Initial instruction coverage: 3380 / 24894
13.58%
Amplification results with 3 amplified tests.
Amplified instruction coverage: 3389 / 24894
13.61%

You see that I used _JacocoCoverageSelector_, because running mutation analysis on your test would be too long, it took almost 33 minutes with JaCoCo which much more faster than mutation analysis.

Could please, update your **DSpot** version, and try on your side? We will compare our outputs.

**WARNING**, since @nicolabertazzo is working on the maven-plugin, the structure of **DSpot** as radically change. There is now a module **dspot** within a module **dspot-parent**.

You may need to change some path of your options/configuration/properties.

Cheers,

-- Benjamin.

**Comment by jesus-gorronogoitia** [comment]
Bonjour Benjamin
I already tested successfully DSpot a week ago, but I forgot to report it to you, sorry.
The results I got were:

> ======= REPORT =======
> Initial instruction coverage: 2936 / 25103
> 11.70%
> Amplification results with 1 amplified tests.
> Amplified instruction coverage: 2945 / 25103
> 11.73%

using the following script.
```
#!/bin/sh
cd /home/yosu/Projects/STAMP/Git/dspot
java -Dis.admin.user=admin -Dis.admin.passwd=$2pRSid# \
-jar target/dspot-1.0.0-jar-with-dependencies.jar \
-p ../../../Supersede/Git/integration/IF/API/eu.supersede.if.api/dspot.properties \
-i 1 -t eu.supersede.integration.api.replan.controller.proxies.test.ReplanControllerProxyTest \
-a MethodAdd \
-s JacocoCoverageSelector
```
Now, I will pull Dspot from GitHub and try again as you recommend

**Comment by monperrus** [comment]
Hi @jesus-gorronogoitia , is the problem fixed? can we close this issue?

**Comment by jesus-gorronogoitia** [comment]
Not yet, I am doing experiments as suggested by Benjamin. I am planning to report asap. Using Clover selector, DSpot gets frozen at certain time with no further explanation (not exception thrown). With Jacoco selector and 3 iterations, yesterday, after 8 hours it was still processing, but with low CPU usage, though at least working, but as I needed to leave the office and I was using my laptop for testing, I stopped the execution. Today, I am running the same experiment in my laptop, but I will be able to keep it running for longer time. Nonetheless, this behavior of low CPU usage deserves to be reported and investigated. Long execution times should be investigated and solved as well. For that, we have T1.4, starting on May, we are leading. We can discuss about this at Oslo.
